### ISSUE TITLE
False positive in `new_ret_no_self` if self type carries lifetime annotation?

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#734]
Example (from [lib_json](https://github.com/valarauca/lib_json)):
```
impl<'a> CharIter<'a> {
    pub fn new<'b>( s: &'b str ) -> CharIter<'b> { .. }
    ..
}
```
While the warning can be fixed by changing it to `pub fn new(s: &'a str) -> CharIter<'a> { .. }`, I am unsure if the lifetime difference is really relevant here.

**Comment by mcarton** [comment]
I should probably not compare types with `==`. I’ll look into that tomorrow.

**Comment by mcarton** [comment]
If I use `ctxt::erase_regions` and `InferCtx::can_equate` instead of simply `==`, it fixes lifetime problems. But I still have a problem with:
```rust
impl<T> CharIter<T> {
    pub fn new<U>( s: &str ) -> CharIter<U> { .. }
    ..
}
```

Any idea how to compare types with type parameters? :sos:

**Comment by llogiq** [comment]
This is a bit complex, because the types can literally come from anywhere. But in general, we should consider two unbound type variables in the same position equal, regardless there name, so

```
impl<T> CharIter<T> {
    pub fn new<U>(..) -> CharIter<U> { .. }
    // should be considered equal to
    pub fn new2(..) -> CharIter<T> { .. }
}
```
Both `T` and `U` are introduced as unbound type variables, one by the inherent impl, the other by the function.

**Comment by mcarton** [comment]
> But in general, we should consider two unbound type variables in the same position equal, regardless there name

I know, but any idea what function does that in rustc?

**Comment by llogiq** [comment]
I don't think there's a function for this in rustc (because for rustc the types are not equal at all). We'll have to walk the type's `TyParam`s and compare their bounds. If the bounds are equal (either unbounded or bounded by some trait)), we should deem the types equal.

**Comment by phansch** [comment]
triage:

The first example is fixed (I'm going to add a testcase for that).
The second example still triggers, while it shouldn't:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6a4da2c0118b2f6dfac4456041c2b332

**Comment by Jarcho** [comment]
This was fixed in #6952
